### PR TITLE
Add http2 support for timelock server

### PR DESCRIPTION
**Pull request by hsaraogi**
---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1594)

### CODE REVIEW

File: atlasdb-api/versions.lock [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+            "locked": "2.8.6",
             "transitive": [
                 "com.fasterxml.jackson.core:jackson-databind",
                 "com.palantir.atlasdb:atlasdb-commons",
                 "com.palantir.atlasdb:timestamp-api"
             ]
         },
         "com.fasterxml.jackson.core:jackson-core": {
-            "locked": "2.6.7",
+            "locked": "2.8.6",
```
Comment (schlosna) [inline atlasdb-api/versions.lock]: I don't think we can bump Jackson above 2.6.x in the `atlasdb-api` or other externally consumed modules as it will cause downstream version conflicts. If we have to bump Jackson for DW, we will have to confine it to the `timelock-server` module.

File: versions.props [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ch.qos.logback:* = 1.1.3
 com.codahale.metrics:metrics-core = 3.0.2
-com.fasterxml.jackson.*:* = 2.6.7
+com.fasterxml.jackson.*:* = 2.8.6
```
Comment (schlosna) [inline versions.props]: See above comment, but we need to find a way for `timelock-server` only to use 2.8.6 and all other modules to use 2.6.7 due to downstream version conflicts.

File: atlasdb-api/versions.lock (hunk 1 above)
Comment (jboreiko) [inline atlasdb-api/versions.lock]: agreed with this, I doubt this would go over well.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java [hunk 3]
```diff
@@ -73,6 +73,8 @@
                             // CipherSuite.TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256,
                             // CipherSuite.TLS_RSA_WITH_AES_256_GCM_SHA384,
                             // CipherSuite.TLS_RSA_WITH_AES_128_GCM_SHA256,
+                            CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
```
Comment (jboreiko) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: For the uninitiated can we get a justification of why we need this cipher suite.  This change was originally made in order to prevent slowness associated with these suites.

File: atlasdb-console/versions.lock [hunk 4]
```diff
@@ -338,13 +347,13 @@
...
         "io.dropwizard:dropwizard-jackson": {
-            "locked": "0.9.3",
+            "locked": "1.0.6",
             "transitive": [
                 "com.palantir.atlasdb:atlasdb-config"
             ]
         },
         "io.dropwizard:dropwizard-util": {
-            "locked": "0.9.3",
+            "locked": "1.0.6",
```
Comment (jboreiko) [inline atlasdb-console/versions.lock]: bumping dropwizard may also cause us problems down stream.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java (hunk 3 above)
Comment (schlosna) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: We should add a comment here referencing https://http2.github.io/http2-spec/index.html#rfc.section.9.2.2

> The black list includes the cipher suite that TLS 1.2 makes mandatory, which means that TLS 1.2 deployments could have non-intersecting sets of permitted cipher suites. To avoid this problem causing TLS handshake failures, deployments of HTTP/2 that use TLS 1.2 MUST support TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 [TLS-ECDHE] with the P-256 elliptic curve [FIPS186].

I know we're also gathering some perf comparisons of throughput and latency, would like to include as much detail in the PR here as possible.

File: atlasdb-console/versions.lock (hunk 4 above)
Comment (schlosna) [inline atlasdb-console/versions.lock]: yeah, I think we have to consolidate the Jackson & Dropwizard version bumps solely to the `timelock-server` module as downstream won't take a direct dependency on it. I don't know how hard it will be to make the current gradle module setup play nice with that, but I'm wondering if we will need to split `timelock-server` out into a separate repo, but defer to those with more gradle-fu than I.

Comment (jeremyk-91) [inline atlasdb-console/versions.lock]: The other projects currently don't depend on `timelock-server`; I do remember it is possible to override the version-locking setup we have. Happy to take a more detailed look at this

Comment (hsaraogi) [inline atlasdb-console/versions.lock]: I am seeing how we can restrict only timelock-server to bump the dropwizard and jackson dependencies. It seems that many of these dependencies are pulled in transitively from other AtlasDB projects. I think we can exclude the transitive ones and add the updated versions only for timelock-server.

File: atlasdb-api/versions.lock (hunk 1 above)
Comment (hsaraogi) [inline atlasdb-api/versions.lock]: This was a do not merge PR, and I thought it would be better if we invest in perf testing before handling in the project.

File: atlasdb-console/versions.lock (hunk 4 above)
Comment (tpetracca) [inline atlasdb-console/versions.lock]: We could always make timelock it's own repo....

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/TextDelegateDecoder.java [hunk 5]
```diff
@@ -59,15 +64,23 @@ public TextDelegateDecoder(Decoder delegate) {
...
+            Collection<String> contentTypes = response.headers().entrySet().stream()
+                    .filter(entry -> entry.getKey().compareToIgnoreCase(HttpHeaders.CONTENT_TYPE) == 0)
+                    .map(Map.Entry::getValue)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toSet());
+            // In the case of multiple content types, or an unknown content type, we'll use the delegate instead.
+            if (contentTypes != null
+                    && contentTypes.size() == 1
+                    && Iterables.getOnlyElement(contentTypes, "").equals(MediaType.TEXT_PLAIN)) {
+                return stringDecoder.decode(response, type);
```
Comment (schlosna) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/TextDelegateDecoder.java]: I assume the root cause here is that this is picking up older version of Feign that does not yet have the fix for https://github.com/OpenFeign/feign/pull/418 . Note that http-remoting ran into this issue, see https://github.com/palantir/http-remoting/pull/156 and specifically its own [implementation of TextDelegateDecoder with case insensitive headers](https://github.com/palantir/http-remoting/pull/156/files#diff-300aea212cd0d3ee78463cf66522f820R45) that ideally we would use or at least keep in sync with.

`com.palantir.atlasdb.keyvalue.remoting.OctetStreamDelegateDecoder` also has same problem with case sensitive headers.

We can avoid doing the collect and instead stream everything:

```
            if (response.headers().entrySet().stream()
                    .filter(entry -> HttpHeaders.CONTENT_TYPE.equalsIgnoreCase(entry.getKey()))
                    .map(Map.Entry::getValue)
                    .flatMap(Collection::stream)
                    .anyMatch(MediaType.TEXT_PLAIN::equals)) {
                return stringDecoder.decode(response, type);
            }
```

Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/TextDelegateDecoder.java]: Ack. I'm a bit concerned about stream usage at all (this is a hot codepath, presumably) but certainly avoiding the collect makes sense. I'll clean this up separately

File: timelock-server/build.gradle [hunk 6]
```diff
@@ -43,11 +45,32 @@ dependencies {
...
 
-    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing'
+    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing', version: libVersions.timelock_dropwizard
 }
 
 distribution {
     serviceName "timelock-server"
     mainClass 'com.palantir.atlasdb.timelock.TimeLockServerLauncher'
     args 'server', 'var/conf/timelock.yml'
+    defaultJvmOpts "-javaagent:service/javaAgents/jetty-alpn-agent.jar"
```
Comment (schlosna) [inline timelock-server/build.gradle]: We probably want to specify a default min & max heap size for timelock. I would suggest we start with `-Xms2g -Xmx2g` based on test runs that implicitly selected `-XX:InitialHeapSize=120634624 -XX:MaxHeapSize=1930153984`.

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/TextDelegateDecoder.java (hunk 5 above)
Comment (schlosna) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/TextDelegateDecoder.java]: We should revert the changes to this file so that we just rebase in the fix from #1613

File: atlasdb-dropwizard-bundle/versions.lock [hunk 7]
```diff
@@ -763,35 +763,36 @@
             ]
         },
         "org.eclipse.jetty:jetty-continuation": {
-            "locked": "9.2.17.v20160517",
+            "locked": "9.4.0.v20161208",
```
Comment (schlosna) [inline atlasdb-dropwizard-bundle/versions.lock]: I think we probably want to pin the Jetty version bump to the `timelock-server` module similar to DW version, as bumping these will likely cause downstream version dependency pains.

File: timelock-server/build.gradle [hunk 8]
```diff
@@ -27,8 +27,14 @@ dependencies {
...
     compile group: 'io.atomix.catalyst', name: 'catalyst-netty'
-    compile group: 'io.dropwizard', name: 'dropwizard-core'
+    compile group: 'io.dropwizard', name: 'dropwizard-core', version: libVersions.timelock_dropwizard
+    compile group: 'io.dropwizard', name: 'dropwizard-http2', version: libVersions.timelock_dropwizard
+    compile group: 'io.dropwizard', name: 'dropwizard-request-logging', version: libVersions.timelock_dropwizard
     compile group: 'io.dropwizard.modules', name: 'dropwizard-java8'
+    compile group: 'org.eclipse.jetty', name: 'jetty-alpn-server'
+    compile group: 'org.eclipse.jetty.http2', name: 'http2-common'
+
+    runtime group: 'org.mortbay.jetty.alpn', name: 'jetty-alpn-agent'
```
Comment (schlosna) [inline timelock-server/build.gradle]: see comment on versions.lock above, but I think we have to pin separate Jetty versions here versus the `atlasdb-dropwizard-bundle`

File: versions.props [hunk 9]
```diff
@@ -34,6 +34,9 @@ org.apache.thrift:libthrift = 0.9.2
 org.assertj:assertj-* = 3.5.2
 org.clojure:clojure = 1.8.0
 org.codehaus.groovy:groovy-all = 2.4.4
+org.eclipse.jetty:* = 9.4.0.v20161208
+org.eclipse.jetty.http2:* = 9.4.0.v20161208
```
Comment (schlosna) [inline versions.props]: see comment on versions.lock above, but I think we have to pin separate Jetty versions for `timelock-server` versus the `atlasdb-dropwizard-bundle`

File: atlasdb-dropwizard-bundle/versions.lock (hunk 7 above)
Comment (hsaraogi) [inline atlasdb-dropwizard-bundle/versions.lock]: We are using the jetty-alpn-agent(https://github.com/jetty-project/jetty-alpn-agent) instead so that the boot jar is loaded as per the JVM version.

File: timelock-server/build.gradle (hunk 6 above)
Comment (schlosna) [inline timelock-server/build.gradle]: We definitely want to set min & max to the same value so that we avoid having to expand the heap as that requires an expensive `Full GC` (see several ~100ms pauses `Full GC (Ergonomics)` below from a recent test run).

```
2017-02-20T23:25:18.043+0000: 11902.330: [Full GC (Ergonomics) [PSYoungGen: 1345K->0K(29696K)] [ParOldGen: 86478K->20319K(112128K)] 87824K->20319K(141824K), [Metaspace: 41836K->41657K(1087488K)], 0.1119925 secs] [Times: user=0.23 sys=0.05, real=0.11 secs]
2017-02-21T10:14:49.331+0000: 50873.617: [Full GC (Ergonomics) [PSYoungGen: 512K->0K(125440K)] [ParOldGen: 112914K->23510K(114688K)] 113426K->23510K(240128K), [Metaspace: 41842K->41842K(1087488K)], 0.0827761 secs] [Times: user=0.18 sys=0.01, real=0.08 secs]
2017-02-21T14:35:47.501+0000: 66531.787: [Full GC (Ergonomics) [PSYoungGen: 1008K->0K(60928K)] [ParOldGen: 114963K->22286K(110592K)] 115972K->22286K(171520K), [Metaspace: 41873K->41873K(1087488K)], 0.0973755 secs] [Times: user=0.14 sys=0.00, real=0.09 secs]
2017-02-21T17:38:56.790+0000: 77521.076: [Full GC (Ergonomics) [PSYoungGen: 1104K->0K(37376K)] [ParOldGen: 110465K->23557K(108544K)] 111569K->23557K(145920K), [Metaspace: 41878K->41878K(1087488K)], 0.0721954 secs] [Times: user=0.14 sys=0.00, real=0.07 secs]
```

GC summary:
```
what           count          total          mean           max     stddev
gen0t           4120         16.801       0.00408         0.023     0.0014
gen1t              6          0.510       0.08496         0.112     0.0268
GCPause         4126         17.311       0.00420         0.112     0.0035
alloc           4120     213947.143      51.92892       170.500    34.4422
promo           4120        343.944       0.08348         7.146     0.1790
used0           4120     219741.247      53.33525       175.068    34.9468
used1              6        434.959      72.49316       112.269    49.4586
used            4126     479652.996     116.25133       266.163    45.0535
commit0         4120     222042.500      53.89381       181.500    35.2669
commit1         4120     430553.500     104.50328       112.500    10.4225
commit          4120     652596.000     158.39709       273.000    39.3772
used0AfterGC    4120       5794.839       1.40651         9.145     0.8145
used1AfterGC       6        114.729      19.12142        23.005     6.2507
usedAfterGC     4126     265714.431      64.40001       113.254    24.7004

alloc/elapsed_time    = 213947.143 MB /  77703.668 s  =   2.753 MB/s
alloc/tot_cpu_time    = 213947.143 MB / 310814.672 s  =   0.688 MB/s
alloc/mut_cpu_time    = 213947.143 MB / 310745.428 s  =   0.688 MB/s
promo/elapsed_time    =    343.944 MB /  77703.668 s  =   0.004 MB/s
promo/gc0_time        =    343.944 MB /     16.801 s  =  20.471 MB/s
gc_seq_load           =     69.244 s  / 310814.672 s  =   0.022%
gc_conc_load          =      0.000 s  / 310814.672 s  =   0.000%
gc_tot_load           =     69.244 s  / 310814.672 s  =   0.022%
```

File: atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java [hunk 10]
```diff
@@ -64,6 +64,13 @@
             new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
                     .tlsVersions(TlsVersion.TLS_1_2)
                     .cipherSuites(
+                            // This GCM cipher suite is for HTTP/2 over TLS1.2 as clients have to
+                            // have at least one permitted cipher suite not in the blacklist. The
+                            // timelock server will support only HTTP/2 connections, and this will ensure
```
Comment (jeremyk-91) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/http/AtlasDbHttpClients.java]: We're not actually preventing the fallback right? So a client that doesn't support HTTP/2 is actually still being supported.

File: timelock-server/build.gradle (hunk 8 above)
Comment (jeremyk-91) [inline timelock-server/build.gradle]: Yep. The ALPN stuff isn't brought in by the atlasdb-dropwizard-bundle at all (it doesn't support HTTP/2 connectors) though, so no separate config is needed here.

File: timelock-server/build.gradle [hunk 11]
```diff
@@ -39,11 +44,16 @@ dependencies {
...
 
-    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing'
+    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing', version: libVersions.timelock_dropwizard
 }
 
 distribution {
     serviceName "timelock-server"
     mainClass 'com.palantir.atlasdb.timelock.TimeLockServerLauncher'
     args 'server', 'var/conf/timelock.yml'
+    defaultJvmOpts "-javaagent:service/lib/jetty-alpn-agent-2.0.6.jar"
```
Comment (jeremyk-91) [inline timelock-server/build.gradle]: Could we extract `2.0.6` to be a constant?

File: atlasdb-config/build.gradle [hunk 12]
```diff
@@ -22,6 +22,7 @@ dependencies {
     compile group: 'com.fasterxml.jackson.dataformat', name: 'jackson-dataformat-yaml'
     compile group: 'io.dropwizard', name: 'dropwizard-jackson'
     compile group: 'com.google.code.findbugs', name: 'annotations'
+    runtime group: 'org.mortbay.jetty.alpn', name: 'jetty-alpn-agent'
```
Comment (jeremyk-91) [inline atlasdb-config/build.gradle]: nit: I'd prefer the runtime dependency to have its own block

File: timelock-server/build.gradle [hunk 13]
```diff
@@ -57,3 +58,27 @@ distribution {
...
+        }
+        forcedModules = [
+                'ch.qos.logback:logback-classic:1.1.7',
+                'com.fasterxml.jackson.module:jackson-module-afterburner:2.6.7',
+                'com.netflix.feign:feign-core:8.10.1',
+                'io.atomix.catalyst:catalyst-transport:1.1.2',
+                'org.objenesis:objenesis:2.3'
+        ]
+    }
+}
```
Comment (jeremyk-91) [inline timelock-server/build.gradle]: nit: newline at end

File: timelock-server/build.gradle [hunk 14]
```diff
@@ -57,3 +58,27 @@ distribution {
...
+            //specifying a fixed version for all libraries with 'io.dropwizard' group
+            if (details.requested.group == 'io.dropwizard') {
+                details.useVersion libVersions.timelock_dropwizard
+            }
+            if (details.requested.group == 'org.eclipse.jetty') {
+                details.useVersion libVersions.timelock_jetty
+            }
+            if (details.requested.group == 'org.glassfish.jersey.core') {
+                details.useVersion libVersions.timelock_jersey
+            }
```
Comment (jeremyk-91) [inline timelock-server/build.gradle]: nit: we could probably string-switch on the groups?

File: docs/source/release_notes/release-notes.rst [hunk 15]
```diff
@@ -40,8 +40,12 @@ develop
...
          - Change
 
-    *    -
-         -
+    *    - |fixed|
+         - Timelock server now supports HTTP/2 and the AtlasDB HTTP clients enable a required GCM cipher suite.
+           This feature improves performance of the Timelock server. Any client that wishes to connect to the timelock server
+           via HTTP2 must add the jetty_alpn_agent as a javaAgent JVM argument, otherwise connections will fall back to
+           HTTP/1.1 and performance may be slow.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/1594>`__)
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: Nicely done! A few suggestions:

- I think this should be `new`; not supporting HTTP/2 doesn't really mean Timelock was broken at all...
- nit: We should standardise `HTTP/2` (line 46)

Also, it would be nice if we could give a concrete example of how to configure the Jetty ALPN agent (i.e. `-javaagent:service/lib/jetty-alpn-agent-2.0.6.jar`). This could be another (small) PR and in a separate section of the docs (and we can include a link here on how to do it).

@rhero can advise better here.

Comment (rhero) [inline docs/source/release_notes/release-notes.rst]: I addressed the release notes nits. Also as an example of how to configure the ALPN agent, I just linked to the line in this PR where we do it for timelock!

File: timelock-server/build.gradle [hunk 16]
```diff
@@ -39,11 +45,44 @@ dependencies {
...
 
-    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing'
+    integTestCompile group: 'io.dropwizard', name: 'dropwizard-testing', version: libVersions.timelock_dropwizard
 }
 
 distribution {
     serviceName "timelock-server"
     mainClass 'com.palantir.atlasdb.timelock.TimeLockServerLauncher'
     args 'server', 'var/conf/timelock.yml'
+    defaultJvmOpts "-Xms512m",
```
Comment (jboreiko) [inline timelock-server/build.gradle]: nit: I'm worried this could have unintended consequences that could be conflated with the http2 change we are attempting to examine.  Wondering if we should delay this heap setting to a separate PR.

File: timelock-server/build.gradle [hunk 17]
```diff
@@ -39,11 +45,44 @@ dependencies {
...
+            "-Xmx512m",
+            "-javaagent:service/lib/jetty-alpn-agent-${libVersions.jetty_alpn_agent}.jar"
+}
+
+dependencyRecommendations {
```
Comment (jboreiko) [inline timelock-server/build.gradle]: nit: I believe this is the default setting so I don't think we need to override it here.  https://github.com/nebula-plugins/nebula-dependency-recommender-plugin#541--conflictresolved-strategy-default

File: timelock-server/build.gradle [hunk 18]
```diff
@@ -39,11 +45,44 @@ dependencies {
...
+    strategy ConflictResolved
+}
+
+configurations.all {
```
Comment (jboreiko) [inline timelock-server/build.gradle]: nit: In general seems like we could move this logic into dependencyRecommendations but if it's working, I don't see much point in changing it now.  The current format is reasonably readable.

File: timelock-server/build.gradle (hunk 17 above)
Comment (hsaraogi) [inline timelock-server/build.gradle]: Its overridden for all projects, so we need to override if we want a different resolution by nebula for this particular module.

File: timelock-server/build.gradle (hunk 18 above)
Comment (hsaraogi) [inline timelock-server/build.gradle]: leaving it as is for now.

File: docs/source/release_notes/release-notes.rst (hunk 15 above)
Comment (hsaraogi) [inline docs/source/release_notes/release-notes.rst]: We should definitely add docs for products to start using the timelock server. I will submit a PR with other doc changes soon.

File: timelock-server/build.gradle (hunk 16 above)
Comment (schlosna) [inline timelock-server/build.gradle]: @jboreiko I don't have strong preference if its in this PR or subsequent, but we should make sure it makes the same release.
